### PR TITLE
When encountering move error on an `Option`, suggest using `as_ref`

### LINKED ISSUE #61109: Suggest `as_ref()` in E0507 if applicable
[Given](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8524415389b142d87a17f9056c4a7aca):

```
pub struct LipogramCorpora {
    selections: Vec<(char, Option<String>)>,
}

impl LipogramCorpora {
    pub fn validate_all(&mut self) -> Result<(), char> {
        for selection in &self.selections {
            if selection.1.is_some() {
                if selection.1.unwrap().contains(selection.0) {
                    return Err(selection.0);
                }
            }
        }
        Ok(())
    }
}
```

`rustc` should suggest using `selection.1..as_ref()`:

```
error[E0507]: cannot move out of borrowed content
 --> src/lib.rs:9:20
  |
9 |                 if selection.1.unwrap().contains(selection.0) {
  |                    ^^^^^^^^^^^
  |                    |
  |                    cannot move out of borrowed content
  |                    help: consider borrowing instead of moving: `selection.1.as_ref()`
```

cc https://github.com/rust-lang/rust/pull/51100, https://github.com/rust-lang/rust/pull/57158

_From https://thenewwazoo.github.io/clone.html_

### PR DESCRIPTION
**Pull request by estebank**
Fix #61109, cc #15457.

### CODE REVIEW

**Comment by czipperz** [comment]
Does this correctly handle `as_mut` as well?

**Comment by czipperz** [comment]
Perhaps this should support `Result` as well?

**Comment by estebank** [comment]
@czipperz adding `Result` support now.

> Does this correctly handle `as_mut` as well?

Could you expand?

**Comment by czipperz** [comment]
I don't know if this is possible, but it would be nice if it suggested `as_mut` when you need mutable access to call a method.  `as_ref().unwrap()` converts `Option<T>` to `&T` but sometimes you need `&mut T`.

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 1]
```diff
@@ -257,81 +252,95 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
...
+                        "{:?}",
+                        original_path.ty(self.mir, self.infcx.tcx).ty,
+                    );
+                    let snippet = self.infcx.tcx.sess.source_map().span_to_snippet(span).unwrap();
+                    let is_option = orig_path_ty.starts_with("std::option::Option");
+                    let is_result = orig_path_ty.starts_with("std::result::Result");
+                    if  is_option || is_result {
+                        err.span_suggestion(
+                            span,
+                            "consider borrowing the `Option`'s content",
```
Comment (czipperz) [inline src/librustc_mir/borrow_check/move_errors.rs]: This is wrong now if it is a `Result`.

**Review by czipperz — CHANGES_REQUESTED** [review]

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 2]
```diff
@@ -257,81 +252,95 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
...
+                                span, &source.to_string(), origin
+                            )
+                        },
+                    };
+                    let orig_path_ty = format!(
+                        "{:?}",
+                        original_path.ty(self.mir, self.infcx.tcx).ty,
+                    );
+                    let snippet = self.infcx.tcx.sess.source_map().span_to_snippet(span).unwrap();
+                    let is_option = orig_path_ty.starts_with("std::option::Option");
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/move_errors.rs]: How do you feel about waiting a bit with merging this PR (a bit being until https://github.com/rust-lang/rust/pull/60966 is merged)? Then you can annotate `Option` and `Result` with `#[rustc_diagnostic_item = "option"]` and `"result"` respectively and then check for these items via

```rust
if let Some(adt) = original_path.ty(self.mir, self.infcx.tcx).adt_def() {
	if self.infcx.tcx.is_diagnostic_item(adt.did, sym::option) {
        ...
    }
}
```

Comment (estebank) [reply src/librustc_mir/borrow_check/move_errors.rs]: There are a couple of places where we're already doing string comparisons like this, so I love the fact that we'll be able to get rid of them soon!

Comment (oli-obk) [reply src/librustc_mir/borrow_check/move_errors.rs]: Hmm... or we can just merge this and fix it up later, doesn't seem like a big deal in fact.

**Comment by estebank** [comment]
@czipperz suggesting both is bit heavy, [but can be done](https://github.com/rust-lang/rust/pull/61147/commits/d2e933d4e65457918f27d3c36feb00008e0feac3). I'm a bit ambivalent on whether it is worth it. On one hand, it teaches the API surface nicely, on the other hand it is verbose and might confuse newcomers.

**Comment by oli-obk** [comment]
My vote is for just suggesting `as_ref` and figuring out how to make sure we emit good errors on subsequent errors that may occur due to the code actually requiring mutable references

File: src/librustc_mir/borrow_check/move_errors.rs (hunk 2 above)
Comment (estebank) [reply src/librustc_mir/borrow_check/move_errors.rs]: I'm fine either way

**Comment by estebank** [comment]
Should I just remove the last commit or are there any further changes needed (beyond #60966)?

**Comment by oli-obk** [comment]
r=me with last commit removed. I agree with your assessment that we'd be overwhelming users for little gain

**Comment by nnethercote** [comment]
https://twitter.com/nnethercote/status/1083507450249043969

**Pull request merged**